mirrord-layer/src/tcp.rs

```diff
@@ -0,0 +1,179 @@
+use std::{
+    collections::HashSet,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+};
+
+use anyhow::Result;
+/// TCP Traffic management, common code for stealing & mirroring
+use async_trait::async_trait;
+use mirrord_protocol::{NewTCPConnection, TCPClose, TCPData};
+use tokio::{
+    net::TcpStream,
+    select,
+    sync::mpsc::{channel, Receiver, Sender},
+};
+use tracing::error;
+
+use crate::{
+    common::Listen,
+    sockets::{SocketInformation, CONNECTION_QUEUE},
+};
+
+const CHANNEL_SIZE: usize = 1024;
+
+#[derive(Debug)]
+pub enum TrafficIn {
+    Listen(Listen),
+    NewConnection(NewTCPConnection),
+    Data(TCPData),
+    Close(TCPClose),
+}
+
+/// To be used by traffic stealer
+pub enum TrafficOut {}
+
+/// Struct responsible for managing the traffic handler
+/// Communicates using incoming/outgoing channels.
+pub struct TCPConfig {
+    /// This is reserved for stealing API.
+    #[allow(dead_code)]
+    outgoing: Sender<TrafficOut>,
+    incoming: Receiver<TrafficIn>,
+}
+
+/// Struct for controlling the traffic stealing struct.
+pub struct TCPApi {
+    outgoing: Sender<TrafficIn>,
+    /// This is reserved for stealing API.
+    #[allow(dead_code)]
+    incoming: Receiver<TrafficOut>,
+}
+
+impl TCPApi {
+    pub async fn send(&self, msg: TrafficIn) -> Result<()> {
+        Ok(self.outgoing.send(msg).await?)
+    }
+
+    /// This is reserved for stealing API.
+    #[allow(dead_code)]
+    pub async fn recv(&mut self) -> Option<TrafficOut> {
+        self.incoming.recv().await
+    }
+
+    pub async fn listen_request(&self, listen: Listen) -> Result<()> {
+        self.send(TrafficIn::Listen(listen)).await
+    }
+
+    pub async fn new_tcp_connection(&self, conn: NewTCPConnection) -> Result<()> {
+        self.send(TrafficIn::NewConnection(conn)).await
+    }
+
+    pub async fn tcp_data(&self, data: TCPData) -> Result<()> {
+        self.send(TrafficIn::Data(data)).await
+    }
+
+    pub async fn tcp_close(&self, close: TCPClose) -> Result<()> {
+        self.send(TrafficIn::Close(close)).await
+    }
+}
+
+#[async_trait]
+pub trait TCPHandler {
+    /// Create new TCP handler communicating using given TCPConfig
+    fn new() -> Self;
+
+    /// Run the TCP Handler, usually as a spawned task.
+    async fn run(mut self, mut config: TCPConfig) -> Result<()>
+    where
+        Self: Sized,
+    {
+        while self.is_running() {
+            select! {
+                msg = config.incoming.recv() => {self.handle_incoming_message(msg).await?;},
+            }
+        }
+        Ok(())
+    }
+
+    /// Should the run loop keep running
+    fn is_running(&mut self) -> bool;
+
+    /// Changes the state so is_running will return False
+    fn stop_running(&mut self);
+
+    fn ports(&mut self) -> &HashSet<Listen>;
+    fn ports_mut(&mut self) -> &mut HashSet<Listen>;
+
```

Comment:
These functions look an awlful lot like `Deref` and `DerefMut` for the type.

Also, `fn ports` should be just `&self`, instead of `&mut self`.

Reply:
That's here to have shared behavior between mirror & steal